File: packages/sandbox/daemon/routes/tools.ts
import {
  fetchToolCatalog,
  type McpEndpoint,
  writeEndpointFile,
  writeToolCatalog,
} from "../tools-catalog";
import { jsonResponse, parseJsonBody } from "./body-parser";

export interface ToolsDeps {
  appRoot: string;
  repoDir: string;
}

/**
 * POST /_sandbox/tools/sync — body `{ url, headers, expiresAt? }` (the run's
 * Virtual MCP endpoint). Writes the endpoint file, then lists the endpoint's
 * tools and writes a JSON Schema catalog under `<repo>/.deco/tools/`.
 * Idempotent; overwrites and prunes stale files. Returns `{ count, tools }`.
 * 502 when the endpoint is unreachable/errors (the endpoint file is still
 * written), 500 when the local write fails.
 */
export function makeToolsSyncHandler(deps: ToolsDeps) {
  return async (req: Request): Promise<Response> => {
    let body: unknown;
    try {
      body = await parseJsonBody(req);
    } catch {
      return jsonResponse({ error: "invalid JSON body" }, 400);
    }

    const mcp = body as Partial<McpEndpoint>;
    if (
      typeof mcp?.url !== "string" ||
      typeof mcp.headers !== "object" ||
      mcp.headers === null
    ) {
      return jsonResponse(
        {
          error: "body must be { url: string, headers: Record<string,string> }",
        },
        400,
      );
    }

    const endpoint = {
      url: mcp.url,
      headers: mcp.headers as Record<string, string>,
      ...(typeof mcp.expiresAt === "number"
        ? { expiresAt: mcp.expiresAt }
        : {}),
    };

    try {
      await writeEndpointFile(endpoint, {
        appRoot: deps.appRoot,
        repoDir: deps.repoDir,
      });
    } catch (err) {
      return jsonResponse(
        { error: err instanceof Error ? err.message : String(err) },
        500,
      );
    }

    let tools: Awaited<ReturnType<typeof fetchToolCatalog>>;
    try {
      tools = await fetchToolCatalog(endpoint);
    } catch (err) {
      return jsonResponse(
        { error: err instanceof Error ? err.message : String(err) },
        502,
      );
    }

    try {
      const result = await writeToolCatalog(tools, {
        appRoot: deps.appRoot,
        repoDir: deps.repoDir,
      });
      return jsonResponse(result);
    } catch (err) {
      return jsonResponse(
        { error: err instanceof Error ? err.message : String(err) },
        500,
      );
    }
  };
}

File: packages/sandbox/daemon-go/internal/routes/tools.go
package routes

import (
	"encoding/json"
	"net/http"
	"net/url"

	"github.com/decocms/studio/sandbox-daemon/internal/httpx"
	"github.com/decocms/studio/sandbox-daemon/internal/toolscatalog"
)

type ToolsDeps struct {
	AppRoot string
	RepoDir string
}

type toolsSyncBody struct {
	URL       string            `json:"url"`
	Headers   map[string]string `json:"headers"`
	ExpiresAt *float64          `json:"expiresAt"`
}

// ToolsSync handles POST /_sandbox/tools/sync — body `{ url, headers,
// expiresAt? }` (the run's Virtual MCP endpoint). Writes the endpoint file,
// then lists the endpoint's tools and writes a JSON Schema catalog under
// `<repo>/.deco/tools/`. Idempotent; overwrites and prunes stale files.
// Returns `{ count, tools }`. 502 when the endpoint is unreachable/errors (the
// endpoint file is still written), 500 when the local write fails.
func ToolsSync(deps ToolsDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var body toolsSyncBody
		raw := json.NewDecoder(r.Body)
		if err := raw.Decode(&body); err != nil {
			httpx.Error(w, 400, "invalid JSON body")
			return
		}
		if body.URL == "" || !isAbsoluteURL(body.URL) || body.Headers == nil {
			httpx.Error(w, 400,
				"body must be { url: string (valid URL), headers: Record<string,string> }")
			return
		}

		ep := toolscatalog.Endpoint{URL: body.URL, Headers: body.Headers}
		if body.ExpiresAt != nil {
			ep.ExpiresAt = int64(*body.ExpiresAt)
		}
		opts := toolscatalog.Opts{AppRoot: deps.AppRoot, RepoDir: deps.RepoDir}

		if _, err := toolscatalog.WriteEndpointFile(ep, opts); err != nil {
			httpx.Error(w, 500, err.Error())
			return
		}

		tools, err := toolscatalog.FetchCatalog(r.Context(), ep)
		if err != nil {
			httpx.Error(w, 502, err.Error())
			return
		}

		count, names, err := toolscatalog.WriteCatalog(tools, opts)
		if err != nil {
			httpx.Error(w, 500, err.Error())
			return
		}
		httpx.JSON(w, 200, map[string]any{"count": count, "tools": names})
	}
}

// isAbsoluteURL mirrors `URL.canParse` — a bare path or an empty scheme is not
// a usable MCP endpoint (the decopilot in-process sentinel used to send one).
func isAbsoluteURL(raw string) bool {
	u, err := url.Parse(raw)
	return err == nil && u.Scheme != "" && u.Host != ""
}
